File: go-bookshelf/handlers_test.go
// handlers_test.go — Integration tests. Sends real HTTP requests through
// the full stack (router → handler → db → Postgres) and checks the responses.
// TDD: write tests first, watch them fail, then build to make them pass.
package main

import (
	"testing"
	"net/http"
	"strings"
	"net/http/httptest"
	"encoding/json"
)

func TestCreateBook(t *testing.T){
	// build the request — assemble the envelope
	body := strings.NewReader(`{"title": "Dune", "author": "Frank Herbert"}`)
	req, err := http.NewRequest("POST", "/books", body)
	if err != nil{
		t.Fatalf("could not create request: %v", err)
	}
	req.Header.Set("Content-Type", "application/json")

	// send the request — handler writes its response into the recorder
	responseRecorder := httptest.NewRecorder()
	handler := setupRouter()
	handler.ServeHTTP(responseRecorder, req)

	// check the result — read the notebook
	if responseRecorder.Code != http.StatusCreated {
		t.Errorf("got status %d, want %d", responseRecorder.Code, http.StatusCreated)
	}

	var got Book
	decoder := json.NewDecoder(responseRecorder.Body)
	err = decoder.Decode(&got)
	if err != nil {
		t.Fatalf("could not decode response: %v", err)
	}

	if got.Title != "Dune" {
		t.Errorf("got title %q, want %q", got.Title, "Dune")
	}
	if got.Author != "Frank Herbert" {
		t.Errorf("got author %q, want %q", got.Author, "Frank Herbert")
	}
	if got.Status != StatusWantToRead {
		t.Errorf("got status %q, want %q", got.Status, StatusWantToRead)
	}
	if got.Id == 0 {
		t.Errorf("expected an id, got 0")
	}
}

func TestGetBooks(t *testing.T) {
	// build the request — GET /books, no body needed
	// send the request — handler writes list of books into the recorder
	// check the result — decode into []Book (a slice, not a single book)
	// verify the list is not empty and contains books we expect
}

func TestGetBook(t *testing.T) {
	// build the request — GET /books/1, no body needed
	// send the request — handler writes one book into the recorder
	// check the result — decode into a single Book
	// verify the id, title, author, status match what we created
}

func TestGetBookNotFound(t *testing.T) {
	// build the request — GET /books/99999, an id that doesn't exist
	// send the request — handler should respond with not found
	// check the result — expect 404 status code
}

func TestUpdateBook(t *testing.T) {
	// build the request — PUT /books/1 with JSON body containing new status
	// send the request — handler updates the book and writes it into the recorder
	// check the result — decode into a Book
	// verify the updated fields changed and the rest stayed the same
}

func TestDeleteBook(t *testing.T) {
	// build the request — DELETE /books/1, no body needed
	// send the request — handler deletes the book
	// check the result — expect 200 or 204 (no content)
	// optionally: GET /books/1 again and expect 404 to confirm it's gone
}

func TestCreateBookNoTitle(t *testing.T) {
	// build the request — POST /books with JSON missing the title field
	// send the request — handler should reject it
	// check the result — expect 400 (bad request)
}

func TestCreateBookInvalidStatus(t *testing.T) {
	// build the request — POST /books with a status that isn't one of the 3 allowed
	// send the request — handler should reject it
	// check the result — expect 400 (bad request)
}

File: ts-bookshelf/test/handlers.test.ts
// handlers.test.ts — Integration tests. Same 8 tests as Go.
// Sends requests through Hono → handler → db → Postgres.
//
// Go tests hardcoded /books/1 — worked because Go ran on a fresh DB.
// Shared DB means IDs are unpredictable, so we capture the id from create
// and use it in get/update/delete.

import { describe, test, expect } from "bun:test"
import "./helpers"
import { sendAndExpect } from "./helpers"

// shared across tests — set by create, used by get/update/delete
let bookId: number

describe("books", () => {

    // POST /books — create a book, capture its id for later tests
    test("create book", async () => {
        const res = await sendAndExpect("POST", "/books", { title: "Dune", author: "Frank Herbert"}, 201)
        const body = await res.json()

        expect(body.title).toBe("Dune")
        expect(body.author).toBe("Frank Herbert")
        expect(body.status).toBe("want to read")
        expect(body.id).toBeGreaterThan(0)

        bookId = body.id
    })

    // GET /books — list all, check non-empty
    test("get books", async () => {
        const res = await sendAndExpect("GET", "/books", undefined, 200)
        const body = await res.json()

        expect(body.length).toBeGreaterThan(0)
    })

    // GET /books/:id — get the book we created
    test("get book", async () => {
        const res = await sendAndExpect("GET", `/books/${bookId}`, undefined, 200)
        const body = await res.json()

        expect(body.id).toBe(bookId)
        expect(body.title).not.toBeEmpty()
        expect(body.author).not.toBeEmpty()
        expect(body.status).not.toBeEmpty()
    })

    // GET /books/99999 — not found
    test("get book not found", async () => {
        await sendAndExpect("GET", "/books/99999", undefined, 404)
    })

    // PUT /books/:id — update the book we created
    test("update book", async () => {
        const res = await sendAndExpect("PUT", `/books/${bookId}`, { title: "Dune", author: "Frank Herbert", status: "reading" }, 200)
        const body = await res.json()

        expect(body.status).toBe("reading")
    })

    // DELETE /books/:id — delete the book we created
    test("delete book", async () => {
        await sendAndExpect("DELETE", `/books/${bookId}`, undefined, 204)
    })

    // POST /books — missing title, check 400
    test("create book no title", async () => {
        await sendAndExpect("POST", "/books", { author: "Frank Herbert" }, 400)
    })

    // POST /books — invalid status, check 400
    test("create book invalid status", async () => {
        await sendAndExpect("POST", "/books", { title: "Dune", author: "Frank Herbert", status: "burned" }, 400)
    })
})
